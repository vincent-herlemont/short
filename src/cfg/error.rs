use std::path::PathBuf;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CfgError {
    #[error("setup not found `{0}`")]
    SetupNotFound(String),
    #[error("project not found `{0:?}`")]
    ProjectNotFound(PathBuf),
    #[error("project `{0:?}` already added")]
    ProjectAlreadyAdded(PathBuf),
    #[error("private env dir not found for `{0}`")]
    PrivateEnvDirNotFound(String),
    #[error("private env dir must be an absolute path `{0:?}` for `{1}`")]
    PrivateEnvDirRelativePath(PathBuf, String),
    #[error("public env dir not found for `{0}`")]
    PublicEnvDirNotFound(String),
    #[error("public env dir already unset for `{0}`")]
    PublicEnvAlreadyUnset(String),
    #[error("private env dir already unset for `{0}`")]
    PrivateEnvAlreadyUnset(String),
    #[error("unknown cfg error")]
    Unknown,
}
