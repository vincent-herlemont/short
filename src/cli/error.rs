use std::path::PathBuf;
use thiserror::Error;

type SetupName = String;

#[derive(Error, Debug)]
pub enum CliError {
    #[error("`{0:?}` not found for `{1}`")]
    EnvDirNotFound(PathBuf, SetupName, #[source] std::io::Error),
    #[error("open editor fail")]
    OpenEditorFail,
    #[error("bad input `{0}` try again")]
    ConfirmBadInputTryAgain(String),
    #[error("you have not allowed to delete var `{0}`:`{1}` in {2}")]
    DeleteVarNowAllowed(String, String, String),
    #[error("env must be sync, please change it manually or run \"short env sync\"")]
    EnvFileMustBeSync,
    #[error("env file `0:?` already exist")]
    EnvFileAlreadyExist(PathBuf),
}
