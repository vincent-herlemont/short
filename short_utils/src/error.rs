//! short specific Error type
//! Allow to embedded multiple error type.
use super::result::Result;
use promptly::ReadlineError;
/// Inspiration from : https://github.com/brson/basic-http-server/blob/1ab052719a88e41822b2955d7d72bf161457d47c/src/main.rs#L468
use serde::export::fmt::Debug;
use serde_yaml;

use std::error::Error as StdError;
use std::fmt::Result as FmtResult;
use std::fmt::{Display, Formatter};
use std::io;
use std::path::StripPrefixError;

use std::string::FromUtf8Error;

pub type ExitStatusCode = Option<i32>;

// TODO: Add PartialEq to Error type.
#[derive(Debug)]
pub enum Error {
    Other(String),
    Wrap(String, Box<Error>),
    Io(io::Error),
    SerdeYaml(serde_yaml::Error),
    Which(which::Error),
    StripPrefixError(StripPrefixError),
    ExitStatus(ExitStatusCode),
    RustyLine(ReadlineError),
    StringUtf8(FromUtf8Error),
}

impl Error {
    /// Create new error
    /// # Example
    /// Usually use [`format!`] as string parameter for pass context information.
    ///
    /// ```
    /// use short_utils::error::Error;
    /// let url = "http://.....";
    /// Error::new(format!("fail to get url {}",url));
    /// ```
    pub fn new<S: AsRef<str>>(msg: S) -> Error {
        Error::Other(String::from(msg.as_ref()))
    }

    /// Create new box error
    pub fn new_box(msg: String) -> Box<Error> {
        Box::new(Error::new(msg))
    }

    pub fn wrap<S: AsRef<str>>(msg: S, err: Error) -> Error {
        Error::Wrap(String::from(msg.as_ref()), Box::new(err))
    }

    pub fn is<P>(&self, predicate: P) -> bool
    where
        P: Fn(&Error) -> bool,
    {
        match self {
            Error::Wrap(_, err) => err.is(predicate),
            err => predicate(err),
        }
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match self {
            Error::Other(msg) => write!(f, "{}", msg),
            Error::Wrap(msg, err) => write!(f, "{} : {}", msg, err),
            Error::Io(err) => write!(f, "{}", err),
            _ => write!(f, "{:?}", self),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        use Error::*;

        match self {
            Io(e) => Some(e),
            SerdeYaml(e) => Some(e),
            Which(_) => None,
            Wrap(_, error) => Some(error),
            Other(_) => None,
            StripPrefixError(e) => Some(e),
            ExitStatus(_) => None,
            RustyLine(e) => Some(e),
            StringUtf8(e) => Some(e),
        }
    }
}

impl From<&'static str> for Error {
    fn from(message: &'static str) -> Error {
        Error::Other(String::from(message))
    }
}

impl From<String> for Error {
    fn from(message: String) -> Error {
        Error::Other(message)
    }
}

impl From<io::Error> for Error {
    fn from(error: io::Error) -> Error {
        Error::Io(error)
    }
}

impl From<serde_yaml::Error> for Error {
    fn from(error: serde_yaml::Error) -> Error {
        Error::SerdeYaml(error)
    }
}

impl From<which::Error> for Error {
    fn from(error: which::Error) -> Error {
        Error::Which(error)
    }
}

impl From<StripPrefixError> for Error {
    fn from(error: StripPrefixError) -> Error {
        Error::StripPrefixError(error)
    }
}

impl From<ReadlineError> for Error {
    fn from(error: ReadlineError) -> Error {
        Error::RustyLine(error)
    }
}

impl From<ExitStatusCode> for Error {
    fn from(exit_status: ExitStatusCode) -> Error {
        Error::ExitStatus(exit_status)
    }
}

impl From<FromUtf8Error> for Error {
    fn from(error: FromUtf8Error) -> Error {
        Error::StringUtf8(error)
    }
}

impl Error {
    pub fn exit_code_eq(&self, code: i32) -> Result<bool> {
        if let Error::ExitStatus(exit_status) = self {
            let cmd_code = exit_status.ok_or(Error::new("Process terminated by signal"))?;
            return Ok(code == cmd_code);
        }
        Err(Error::new("no exit status")).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::Error;
    use crate::result::Result;
    use std::io;
    use std::io::ErrorKind;

    #[test]
    fn error_test() {
        fn create_io_error() -> Result<()> {
            Err(Error::from(io::Error::new(io::ErrorKind::Other, "ho no !")))
        }
        fn create_other_error() -> Result<()> {
            Err(Error::new(format!("ho no !")))
        }
        fn create_other_error_from() -> Result<()> {
            Err(Error::from("ho no !"))
        }

        let result = create_io_error();
        match result.err().unwrap() {
            Error::Io(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error_from();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };
    }

    #[test]
    fn nested_error() {
        let err = Error::from(io::Error::new(ErrorKind::NotFound, "not found"));
        let err = Error::wrap("wrap lvl 1", err);
        let err = Error::wrap("wrap lvl 2", err);
        let b = err.is(|err| matches!(err, Error::Io(_)));
        assert!(b);
        let b = err.is(|err| matches!(err, Error::Other(_)));
        assert!(!b);
    }
}
