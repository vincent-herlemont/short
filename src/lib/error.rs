//! D4d specific type of error

use serde::export::fmt::Debug;
use std::error::Error;
use std::fmt::{Display, Formatter, Result};

#[derive(Debug)]
pub struct D4dError {
    pub detail: String,
}

impl D4dError {
    /// Create new error
    /// # Example
    /// Usually use [`format!`] as string parameter for pass context information.
    ///
    /// ```
    /// let url = "http://.....";
    /// D4dError::new(format!("fail to get url",url);
    /// ```
    pub fn new(msg: String) -> D4dError {
        D4dError { detail: msg }
    }

    /// Create new box error
    pub fn new_box(msg: String) -> Box<D4dError> {
        Box::new(D4dError::new(msg))
    }
}

impl Display for D4dError {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(f, "{}", self.detail)
    }
}

impl Error for D4dError {}
