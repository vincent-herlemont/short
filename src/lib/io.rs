//! Input/Ouput manipulation operations related of d4d domain.
use crate::d4d_error;

use std::error::Error;
use std::io::BufRead;

/// Search for a buffer a line that satisfies a predicate
/// Return [`String`] that hold the contain of buffer.
pub fn read_to_string_finds<B, F>(mut buf_read: B, f: F) -> Result<String, Box<dyn Error>>
where
    B: BufRead,
    F: Fn(&str) -> bool,
{
    let mut buf = String::new();
    let mut matched = false;
    while buf_read.read_line(&mut buf)? > 0 as usize {
        if !matched {
            matched = f((&buf).as_str())
        }
    }
    match matched {
        true => Ok(buf),
        false => Err(Box::new(d4d_error::new("fail to match"))),
    }
}

#[cfg(test)]
mod tests {
    use crate::lib;
    use std::io::Cursor;

    #[test]
    fn read_to_string_finds_test() {
        let data = b"line1\nline2\n\ntest";
        let cursor = Cursor::new(data);
        let contents = lib::io::read_to_string_finds(cursor, |s| s.contains("test")).unwrap();
        assert_eq!(contents.into_bytes(), data);
    }

    #[test]
    fn read_to_string_finds_test_fail() {
        let data = b"line1\nline2\n\ntest";
        let cursor = Cursor::new(data);
        let result = lib::io::read_to_string_finds(cursor, |s| s.contains("notfouned"));
        assert!(result.is_err());
    }
}
