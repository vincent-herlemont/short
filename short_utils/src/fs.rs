//! File manipulation operations related of short domain.
use super::io;
use crate::error::Error;
use crate::result::{unwrap_partition, Result};
use std::cmp::Ordering;
use std::fs::File;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`PathBuf`] of files and his content.
#[derive(Debug, Eq, PartialEq)]
pub struct ContentFile {
    pub path: PathBuf,
    pub contents: String,
}

impl Ord for ContentFile {
    fn cmp(&self, other: &Self) -> Ordering {
        self.path.cmp(&other.path)
    }
}

impl PartialOrd for ContentFile {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.path.cmp(&other.path))
    }
}

impl ContentFile {
    /// Return [`ContentFile`] when the predicate [`next_line`] is satisfied.
    pub fn read_contain<F>(path: &PathBuf, next_line: F) -> Result<ContentFile>
    where
        F: Fn(&str) -> bool,
    {
        let contents = match File::open(path).map(|file| BufReader::new(file)) {
            Ok(buffer) => io::read_to_string_contain(buffer, next_line),
            Err(error) => Err(Error::from(error)),
        };

        match contents {
            Ok(contents) => Ok(ContentFile {
                path: path.to_owned(),
                contents,
            }),
            Err(error) => Err(error),
        }
    }

    /// Return each [`ContentFile`] and [`Error`] when the predicate [`next_line`] is satisfied or not.
    pub fn read_contain_multi<F>(paths: &[PathBuf], next_line: F) -> (Vec<ContentFile>, Vec<Error>)
    where
        F: Fn(&str) -> bool,
    {
        let results: (Vec<_>, Vec<_>) = paths
            .into_iter()
            .map(|path| ContentFile::read_contain(path, &next_line))
            .partition(Result::is_ok);
        unwrap_partition(results)
    }
}

#[cfg(test)]
mod tests {
    use crate::asset::default_assets;
    use crate::assets::get_all;
    use crate::error::Error;
    use crate::fs::ContentFile;
    use crate::path::retrieve;
    use crate::test::before;
    use std::io;

    #[allow(unreachable_patterns)]
    #[test]
    fn read_contain_multi_test() {
        let config = before("search_test", default_assets(get_all()));
        let paths = retrieve(&config.tmp_dir).unwrap();
        let (mut content_files, errors) =
            ContentFile::read_contain_multi(&paths, |line| line.contains("test"));

        content_files.sort();
        // ContentFiles
        assert!((&content_files).len() >= 2);
        assert_find!(
            content_files,
            ContentFile { contents, path },
            contents == &String::from("console.log(\'test.js\');")
                && path.to_string_lossy().contains("test.js")
        );

        assert!((&errors).len() > 3);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::Other);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::NotFound);
    }
}
