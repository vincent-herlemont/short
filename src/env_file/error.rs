use std::io;
use std::path::PathBuf;

use thiserror::Error;

#[derive(Error, Debug)]
pub enum EnvReaderError {
    #[error("io env reader error")]
    Io {
        #[from]
        source: io::Error,
    },
    #[error("space on var name `{0}`")]
    SpaceOnVarName(String),
    #[error("unknown env error")]
    Unknown,
}

#[derive(Error, Debug)]
pub enum EnvError {
    #[error("io env error")]
    Io {
        #[from]
        source: io::Error,
    },
    #[error("fs_extra env error")]
    FsExtra {
        #[from]
        source: fs_extra::error::Error,
    },
    #[error("fail to parse `{file:?}`")]
    FailToParse {
        #[source]
        source: EnvReaderError,
        file: PathBuf,
    },
    #[error("env var `{0}` not found in `{1:?}`")]
    EnvVarNotFound(String, PathBuf),
    #[error("env file `{0:?}` has no file name")]
    EnvFileHasNoFileName(PathBuf),
    #[error("env file `{0:?}` has an empty file name")]
    EnvFileNameIsEmpty(PathBuf),
    #[error("env file `{0:?}` has incorrect file name : it must begin with `.` char")]
    EnvFileNameIncorrect(PathBuf),
}
