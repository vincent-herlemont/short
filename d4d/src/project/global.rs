use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File, OpenOptions};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CURRENT_FILE_NAME: &'static str = "projects.yaml";

fn global_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_global_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let path = global_directory_path(home);
    match create_dir(path) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn global_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    global_directory_path(home).join(PROJECT_CURRENT_FILE_NAME)
}

#[allow(dead_code)]
fn read_global_file<P: AsRef<Path>>(home: P) -> Result<GlobalProjects> {
    let path = global_file_path(home);
    let file = File::open(path)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

/// Create or overwrite project config file.
#[allow(dead_code)]
fn save_global_file<P: AsRef<Path>>(home: P, projects: &GlobalProjects) -> Result<()> {
    let path = global_file_path(home);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &projects)?;
    Ok(())
}

#[derive(Serialize, Deserialize, Debug)]
struct GlobalProject {
    name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    path: Option<PathBuf>,

    #[serde(skip_serializing_if = "Option::is_none")]
    current_env: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    git_secret_repo: Option<String>,
}

impl GlobalProject {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> GlobalProject {
        GlobalProject {
            name: String::from(name.as_ref()),
            path: None,
            current_env: None,
            git_secret_repo: None,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GlobalProjects {
    #[serde(skip)]
    home_dir: PathBuf,

    #[serde(rename = "projects")]
    all: Vec<GlobalProject>,
}

impl GlobalProjects {
    pub fn new<P: AsRef<Path>>(home_dir: P) -> Result<GlobalProjects> {
        let home_dir = home_dir.as_ref().to_path_buf();
        match read_global_file(&home_dir) {
            Ok(global_file) => Ok(global_file),
            Err(_) => {
                let global_directory = global_directory_path(&home_dir);
                if !global_directory.exists() {
                    match create_global_directory(&home_dir) {
                        Ok(_) => (),
                        Err(err) => {
                            return Err(Error::wrap(
                                format!(
                                    "fail to create configuration directory {}",
                                    global_directory.to_string_lossy()
                                ),
                                err,
                            ));
                        }
                    }
                }
                // TODO: match for create err only if file does not exist.
                let global_projects = GlobalProjects {
                    home_dir: home_dir.to_owned(),
                    all: vec![],
                };
                match save_global_file(&home_dir, &global_projects) {
                    Ok(_) => Ok(global_projects),
                    Err(err) => Err(Error::wrap(
                        format!(
                            "fail to create global file to {}",
                            home_dir.to_string_lossy()
                        ),
                        err,
                    )),
                }
            }
        }
    }

    pub fn add<N, P>(&mut self, name: N, path: P) -> Result<()>
    where
        N: AsRef<str>,
        P: AsRef<Path>,
    {
        let path = PathBuf::from(path.as_ref())
            .canonicalize()
            .map_err(|e| Error::wrap("fail to get absolute path of : {}", Error::from(e)))?;

        self.all.push(GlobalProject {
            name: String::from(name.as_ref()),
            path: Some(path),
            current_env: None,
            git_secret_repo: None,
        });

        if let Err(err) = save_global_file(&self.home_dir, self) {
            Err(Error::wrap(
                format!(
                    "fail to save global file to : {}",
                    self.home_dir.to_string_lossy(),
                ),
                err,
            ))
        } else {
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::project::global::{
        create_global_directory, global_file_path, read_global_file, save_global_file,
        GlobalProject, GlobalProjects,
    };
    use insta::assert_debug_snapshot;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use std::path::PathBuf;
    use utils::asset::Assets;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_save_global_file() {
        let config = before("test_save_global_file", Assets::All(HashMap::new()));
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![GlobalProject::new("project_1")],
        };
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects:\n  - name: project_1"));

        // Overwrite
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![],
        };
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects: []"));
    }

    #[test]
    fn test_empty_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_empty_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects:
    - name: project_1
    - name: project_2
      current_env: dev
    - name: project_3
      path: /todo/plop/
      current_env: dev
    - name: project_4
      current_env: prod
      git_secret_repo: "git@privategit.com" 
"#,
        );
        let config = before("test_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_global_directory() {
        let config = before("create_global_directory", Assets::All(HashMap::new()));
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_eq!(r.iter().count(), 2);
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_err());
    }
}
