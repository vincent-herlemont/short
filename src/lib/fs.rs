//! File manipulation operations related of d4d domain.
use super::io;
use std::fs::File;
use std::io as sio;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`File`] and his content.
#[derive(Debug)]
struct ContentFile {
    path: PathBuf,
    contents: String,
}

fn read_to_string_finds<F>(paths: &[PathBuf], f: F) -> (Vec<ContentFile>, Vec<sio::Error>)
where
    F: Fn(&str) -> bool,
{
    let (content_files, errors): (Vec<_>, Vec<_>) = paths
        .into_iter()
        .map(|path| {
            let contents = match File::open(path).map(|file| BufReader::new(file)) {
                Ok(buffer) => io::read_to_string_finds(buffer, &f),
                Err(error) => Err(error),
            };

            match contents {
                Ok(contents) => Ok(ContentFile {
                    path: path.to_owned(),
                    contents,
                }),
                Err(error) => Err(error),
            }
        })
        .partition(Result::is_ok);

    (
        content_files.into_iter().map(Result::unwrap).collect(),
        errors.into_iter().map(Result::unwrap_err).collect(),
    )
}

#[cfg(test)]
mod tests {
    use crate::lib::fs::read_to_string_finds;
    use crate::lib::path::retrieve;
    use crate::lib::test::{after, before};

    #[test]
    fn read_to_string_finds_test() {
        let tmpath = before("search_test");
        let paths = retrieve(&tmpath).unwrap();
        let (content_files, errors) = read_to_string_finds(&paths, |line| line.contains("test"));

        // ContentFiles
        assert_eq!((&content_files).len(), 1);
        assert_eq!((&content_files)[0].contents, "console.log(\'test.js\');");
        assert!((&content_files)[0]
            .path
            .to_str()
            .unwrap()
            .contains("test.js"));

        // Errors
        assert_eq!((&errors).len(), 2);
        after(tmpath);
    }
}
