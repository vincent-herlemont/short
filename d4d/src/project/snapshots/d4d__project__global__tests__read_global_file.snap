---
source: d4d/src/project/global.rs
expression: projects
---
Ok(
    GlobalProjects {
        home_dir: "",
        all: [
            GlobalProject {
                name: "project_1",
                path: None,
                current_env: None,
                git_secret_repo: None,
            },
            GlobalProject {
                name: "project_2",
                path: None,
                current_env: Some(
                    "dev",
                ),
                git_secret_repo: None,
            },
            GlobalProject {
                name: "project_3",
                path: Some(
                    "/todo/plop/",
                ),
                current_env: Some(
                    "dev",
                ),
                git_secret_repo: None,
            },
            GlobalProject {
                name: "project_4",
                path: None,
                current_env: Some(
                    "prod",
                ),
                git_secret_repo: Some(
                    "git@privategit.com",
                ),
            },
        ],
    },
)
